### ISSUE TITLE
Proxying socketio through a flask app

### ISSUE THREAD

**Issue by kyouko-taiga** [miguelgrinberg/Flask-SocketIO#423]
I was wondering what would be the best option to proxy another socketio server through a flask app with flask-socketio. The reason I need to achieve this is that my app should act as a bridge between my end users and other socketio servers that run on the backend.

I was considering starting background threads on the server-side that would listen to the backend socketio servers (for instance using [socketIO-client](https://github.com/invisibleroads/socketIO-client)) and emitting messages to my end users whenever the threads receives something. Is there a better approach?

**Comment by miguelgrinberg** [comment]
I guess you could use the Python Socket.IO client (not something I've done, it's on pypi) and act as a proxy, but why not put an nginx process to do this for you, in a much more efficient way?

**Comment by kyouko-taiga** [comment]
Ideally I'd like to perform check for authentication credentials in my "bridge". This socket.io channel corresponds to a shell that is mounted in a docker container (trying another approach to what failed for me in #418).

I could manage to do it with socketIO-client though. For anyone who may be interested (trying to omit as much noise as possible):

```python
from flask_socketio import SocketIO
from socketIO_client import SocketIO as SocketIOClient

# ...

inbound_socket = SocketIO(app)
server_sockets = {}

@inbound_socket.on('connect')
def handle_conect():
    global server_sockets

    # Create a socketio client that'll listen to the remote server.
    outbound_socket = SocketIOClient('localhost', 5000)
    sid = request.sid

    # Forward the data from the remote server to its client.
    def forward_outbound(data):
        print('<-', data, '<-')
        inbound_socket.emit('event', data, room=sid)

    # Define a function that'll keep the socket listening.
    def listing_outbound():
        outbound_socket.on('event', forward_outbound)
        outbound_socket.wait()

    pool.spawn(listing_outbound)
    server_sockets[request.sid] = outbound_socket

@inbound_socket.on('disconnect')
def handle_disconnect():
    global server_sockets

    # Delete the socketio client.
    if request.sid in server_sockets:
        del server_sockets[request.sid]

@inbound_socket.on('event')
def forward_inbound(data):
    if request.sid in server_sockets:
        # Forward the data from the client to its server.
        print('->', data, '->')
        server_sockets[request.sid].emit('event', data)
```

**Comment by jwg4** [comment]
This looks like a nice piece of code, but if you want to think again about using nginx for this, you should be aware that it probably can deal with various types of authentication for you.

**Comment by kyouko-taiga** [comment]
The problem is that what I actually need is to compute the address of the redirection, based on some account information. I could simply give a token to the client that would contain the actual address of the shell endpoint, but I wanted to avoid this approach, as it would have let any authenticated client access any shell (i.e. other's).

I ended up using [Lapis](http://leafo.net/lapis/) (a Lua framework) to create routes within nginx that does just that. I now give a signed token to the client that my Lua route can check before letting nginx redirect the connection. While this is a bit convoluted, this works quite well. If someone was to try this approach, I would recommend [OpenResty](https://openresty.org/en/) rather than Lapis though.

The code of this approach can be found here: https://github.com/umiushi/umiushi/
Unfortunately I had to work on other projects lately, and hadn't had time to clean up this repo, so it is a it messy right now.
